### PR TITLE
[Doc] Readme revamp for efficiency/modularity display

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Gives code snippets of each of the features of the library.
Makes README.md more compelling on the efficiency and modularity standpoints.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+`TensorDict` makes it easy to re-use pieces of code across environments, models and
+algorithms. For instance, here's how to code a rollout in TorchRL:
+    <details>
+      <summary>Code</summary>
+    
+    ```python
+    tensordict = env.reset()
+    policy = TensorDictModule(
+        model, 
+        in_keys=["observation_pixels", "observation_vector"],
```
Comment (msaroufim) [inline README.md:40]: Quick sketch would be useful here

File: README.md [hunk 2]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+        total_frames=10000,
+        frames_per_batch=50,
+        ...
+    )
+    for i, tensordict_data in enumerate(collector):
+        loss = loss_module(tensordict_data)
+        loss.backward()
+        optim.step()
+        optim.zero_grad()
+        collector.update_policy_weights_()
```
Comment (msaroufim) [inline README.md:73]: Outside of changing the collector policy it seems like this is the standard pytorch training loop which was really nice. So a quick picture again would be really helpful because you're basically showing how similar RL is to standard supervised learning but where you change the data loader on the fly

File: README.md [hunk 3]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
-- various tools for distributed learning (e.g. [memory mapped tensors](torchrl/data/tensordict/memmap.py));
-- various [architectures](torchrl/modules/models/) and models (e.g. [actor-critic](torchrl/modules/tensordict_module/actors.py));
-- [exploration wrappers](torchrl/modules/tensordict_module/exploration.py) and [modules](torchrl/modules/models/exploration.py);
+- multiprocess [data collectors](torchrl/collectors/collectors.py)<sup>(2)</sup> that work synchronously or asynchronously:
+    <details>
+      <summary>Code</summary>
+    
+    ```python
+    collector = MultiaSyncDataCollector(
+        [make_env, make_env],
```
Comment (msaroufim) [inline README.md:61]: what's make_env? Worth highlighting a bit how to support your own environments

File: README.md [hunk 4]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+        scratch_dir="/tmp/"
+    )
+    buffer = TensorDictPrioritizedReplayBuffer(
+        buffer_size=10000,
+        alpha=0.7,
+        beta=0.5,
+        collate_fn=lambda x: x,
+        pin_memory=device != torch.device("cpu"),
+        prefetch=10,  # multi-threaded sampling
+        storage=storage
```
Comment (msaroufim) [inline README.md:93]: Mention the interface for storage briefly, for your modularity pitch it may make sense for people to want to creat their own modules

File: README.md [hunk 5]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+from common libraries (OpenAI gym, deepmind control lab, etc.)<sup>(1)</sup> and [wrappers](torchrl/envs/vec_env.py) for parallel execution<sup>(2)</sup>, 
+as well as a new pytorch-first class of [tensor-specification class](torchrl/data/tensor_specs.py):
+    <details>
+      <summary>Code</summary>
+    
+    ```python
+    env_make = lambda: GymEnv("Pendulum-v1", from_pixels=True)
+    env_parallel = ParallelEnv(4, env_make)  # creates 4 envs in parallel
+    tensordict = env_parallel.rollout(max_steps=20)
+    assert tensordict.shape == [4, 20]  # 4 envs, 20 steps rollout
```
Comment (msaroufim) [inline README.md:108]: Love this

File: README.md [hunk 6]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+    ```
+    </details>
+
+- cross-library [environment transforms](torchrl/envs/transforms/transforms.py)<sup>(1)</sup>, 
+executed on device and in a vectorized fashion<sup>(2)</sup>,
```
Comment (msaroufim) [inline README.md:113]: By vectorized fashion do you mean the SIMD acceleration in torchvision? I would just say that

File: README.md [hunk 7]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+    env = TransformedEnv(
+        env_base, 
+        Compose(ToTensorImage(), ObservationNorm(loc=0.5, scale=1.0)),  # executes the transforms once and on device
+    )
+    tensordict = env.reset()
+    assert tensordict.device == torch.device("cuda:0")
+    ```
+    </details>
+
+- various tools for distributed learning (e.g. [memory mapped tensors](torchrl/data/tensordict/memmap.py))<sup>(2)</sup>;
```
Comment (msaroufim) [inline README.md:130]: You can fold this point into the previous point around replay buffers. Have a picture for the data parallel rollouts with the central store

File: README.md [hunk 8]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+- various [architectures](torchrl/modules/models/) and models (e.g. [actor-critic](torchrl/modules/tensordict_module/actors.py))<sup>(1)</sup>:
```
Comment (msaroufim) [inline README.md:131]: Mention a few more, maybe in a github issue describe a few good issues for other models people can contribute

File: README.md [hunk 9]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+        num_cells=[32, 64, 64],
+        kernel_sizes=[8, 4, 3],
+        strides=[4, 2, 1],
+    )
+    common_module = TensorDictModule(
+        common_module,
+        in_keys=["pixels"],
+        out_keys=["hidden"],
+    )
+    policy_module = NormalParamsWrapper(
```
Comment (msaroufim) [inline README.md:148]: What's the goal of the param wrapper?

File: README.md [hunk 10]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+which process and prepare the data coming out of the environments to be used by the agent:
+    <details>
+      <summary>Code</summary>
+    
+    ```python
+    env_make = lambda: GymEnv("Pendulum-v1", from_pixels=True)
+    env_base = ParallelEnv(4, env_make, device="cuda:0")  # creates 4 envs in parallel
+    env = TransformedEnv(
+        env_base, 
+        Compose(ToTensorImage(), ObservationNorm(loc=0.5, scale=1.0)),  # executes the transforms once and on device
```
Comment (msaroufim) [inline README.md:123]: Are there non-pixel-based transforms that would be relevant to showcase?

File: README.md [hunk 11]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+        out_keys=["hidden"],
+    )
+    policy_module = NormalParamsWrapper(
+        MLP(
+            num_cells=[64, 64],
+            out_features=32,
+            activation=nn.ELU,
+        )
+    )
+    policy_module = ProbabilisticTensorDict(  # stochastic policy
```
Comment (msaroufim) [inline README.md:155]: This could be worth elaborating on? So it seems like a tensor dict can take as a parameter another tensordict which helps with your modularity story

File: README.md [hunk 12]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
+    with set_exploration_mode("random"):
+        tensordict = policy_explore(tensordict)  # will use eps-greedy
+    with set_exploration_mode("mode"):
+        tensordict = policy_explore(tensordict)  # will not use eps-greedy
+    ```
+    </details>
+
 - various [recipes](torchrl/trainers/helpers/models.py) to build models that correspond to the environment being deployed;
-- a generic [trainer class](torchrl/trainers/trainers.py).
+- a generic [trainer class](torchrl/trainers/trainers.py)<sup>(1)</sup>.
```
Comment (msaroufim) [inline README.md:190]: bring this near the top with the custom training loop, either build your own loop or use the prebuilt one

**Review by msaroufim — CHANGES_REQUESTED** [review]
Left a bunch of nits but at a high level what I got from the rewrite was I understood why you were excited about the TensorDict abstraction. But I think it's worth elaborating on it even more by discussing some application with a figure and then showing how to implement it with a simple TensorDict, you already do this for example when you discuss the rollouts.

I think the benefit of the rewrite is ultimately making it clear to users which new powerful abstractions they should know about and which specific implementations would you like them to author so make that call to action clear with a link to creating a PR or github issue.

I will say when I saw the custom training loop and how the only difference between that and a regular pytorch training loop is changing the data collector on the fly I felt like it gave me a different simple way to understand RL.

File: README.md [hunk 13]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
 
-It provides pytorch and python-first, low and high level abstractions for RL that are intended to be efficient, documented and properly tested. 
+It provides pytorch and **python-first**, low and high level abstractions for RL that are intended to be **efficient**, **modular**, **documented** and properly **tested**. 
 The code is aimed at supporting research in RL. Most of it is written in python in a highly modular way, such that researchers can easily swap components, transform them or write new ones with little effort.
 
 This repo attempts to align with the existing pytorch ecosystem libraries in that it has a dataset pillar ([torchrl/envs](torchrl/envs)), [transforms](torchrl/envs/transforms), [models](torchrl/modules), data utilities (e.g. collectors and containers), etc. 
 TorchRL aims at having as few dependencies as possible (python standard library, numpy and pytorch). Common environment libraries (e.g. OpenAI gym) are only optional.
 
 On the low-level end, torchrl comes with a set of highly re-usable functionals for [cost functions](torchrl/objectives/costs), [returns](torchrl/objectives/returns) and data processing.
```
Comment (Benjamin-eecs) [inline README.md:21]: Worth mentioning the pure functional returns in https://github.com/facebookresearch/rl/blob/main/torchrl/objectives/returns/functional.py, and the difference between `loss_module` and pure functional returns, maybe not here, but somewhere else. I think the clarification is important.

File: README.md [hunk 14]
```diff
@@ -240,9 +240,18 @@ pip3 install --pre torch torchvision --extra-index-url https://download.pytorch.
 
 and functorch
 ```
+pip3 install ninja  # Makes the build go faster
```
Comment (shagunsodhani) [inline README.md:243]: Maybe we should stick to either `pip` or `pip3` everywhere.

File: README.md [hunk 15]
```diff
@@ -240,9 +240,18 @@ pip3 install --pre torch torchvision --extra-index-url https://download.pytorch.
...
 pip install "git+https://github.com/pytorch/functorch.git"
 ```
 
+If this fails, you can get the latest version of functorch that was marked to be 
+compatible with the current torch version:
+```bash
+pip3 install ninja  # Makes the build go faster
+PYTORCH_VERSION=`python -c "import torch.version; print(torch.version.git_version)"`
+pip install "git+https://github.com/pytorch/pytorch.git@$PYTORCH_VERSION#subdirectory=functorch"
```
Comment (shagunsodhani) [inline README.md:252]: Ohh this is sweet!

Would it also work for the older functorch (0.1.0) ?

**Review by shagunsodhani — APPROVED** [review]

File: README.md (hunk 15 above)
Comment (vmoens) [reply README.md:252]: so basically it's gonna install the functorch that corresponds to your pytorch (so you don't pick up the functorch version but the pytorch version picks up the functorch version -- this is an ancient Chinese proverb by the way)

I'm not sure if that will work with functorch versions that are prior to functorch migration to pytorch/pytorch though

File: README.md (hunk 14 above)
Comment (vmoens) [reply README.md:243]: good point

File: README.md [hunk 16]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
...
 This repo attempts to align with the existing pytorch ecosystem libraries in that it has a dataset pillar ([torchrl/envs](torchrl/envs)), [transforms](torchrl/envs/transforms), [models](torchrl/modules), data utilities (e.g. collectors and containers), etc. 
 TorchRL aims at having as few dependencies as possible (python standard library, numpy and pytorch). Common environment libraries (e.g. OpenAI gym) are only optional.
 
 On the low-level end, torchrl comes with a set of highly re-usable functionals for [cost functions](torchrl/objectives/costs), [returns](torchrl/objectives/returns) and data processing.
 
-On the high-level end, torchrl provides:
-- multiprocess [data collectors](torchrl/collectors/collectors.py);
-- efficient and generic [replay buffers](torchrl/data/replay_buffers/replay_buffers.py);
-- [TensorDict](torchrl/data/tensordict/tensordict.py), a convenient data structure to pass data from one object to another without friction;
+TorchRL aims at a high modularity (1) and good runtime performance (2).
```
Comment (shagunsodhani) [inline README.md:23]: The more common usage is "(1) high modularity and (2) good runtime performance"

File: README.md [hunk 17]
```diff
@@ -10,28 +10,192 @@ The features are available before an official release so that users and collabor
...
+On the high-level end, TorchRL provides:
+- [`TensorDict`](torchrl/data/tensordict/tensordict.py), 
+a convenient data structure<sup>(1)</sup> to pass data from 
+one object to another without friction.
+`TensorDict` makes it easy to re-use pieces of code across environments, models and
+algorithms. For instance, here's how to code a rollout in TorchRL:
+    <details>
+      <summary>Code</summary>
+    
+    ```diff
```
Comment (vmoens) [inline README.md:36]: @msaroufim is this what you had in mind? i.e. showing what the primitives bring via diff code style?

File: README.md (hunk 11 above)
Comment (vmoens) [reply README.md:161]: `ProbabilisticTensorDict` should read `ProbabilisticTensorDictModule`

File: README.md [hunk 18]
```diff
@@ -10,28 +10,186 @@ The features are available before an official release so that users and collabor
 
 ---
 
-TorchRL is an open-source Reinforcement Learning (RL) library for PyTorch. 
+**TorchRL** is an open-source Reinforcement Learning (RL) library for PyTorch. 
 
-It provides pytorch and python-first, low and high level abstractions for RL that are intended to be efficient, documented and properly tested. 
+It provides pytorch and **python-first**, low and high level abstractions for RL that are intended to be **efficient**, **modular**, **documented** and properly **tested**.
```
[the remark this answers is not in the document]
Comment (vmoens) [reply README.md:15]: > On the documented and tested front it may be better to just show instead of tell, have a badge to show your test coverage and link to a high quality hosted documentation. As you build out more content and tutorials make sure to also link to them

Definitely
We already have a badge for the tests (passing or not), but not for coverage.
For the doc, thing is that right now the doc can't sit anywhere: readthedocs is not fb-approved, building a domaing just for the doc should be weighted against waiting for torchrl to migratet to github.com/pytorch: at this point the doc will be integrated in the general doc. Hence we're sitting on our hands for now. However, most functions and classes have an well documented docstring, so it's useful IMO to acknowledge that in the readme for now.
